### PR TITLE
common update mechanism

### PR DESCRIPTION
**Pull request by garbas**
i've took the update script from @zimbatm for servo and generalized it for potentially any package.

- to update one package run ``nix-shell update.nix -A gecko``
- to update all packages run ``nix-shell update.nix``
- also nixpkgs is tracked, to update nixpkgs run ``nix-shell update.nix -A nixpkgs``

### CODE REVIEW

File: lib/update.nix [hunk 1]
```diff
@@ -0,0 +1,57 @@
+{ pkgs_mozilla }:
+
+let
+  inherit (pkgs_mozilla.nixpkgs) cacert nix-prefetch-scripts jq;
+in {
```
Comment (zimbatm) [inline lib/update.nix]: For me the ./lib contains pure functions that don't depend on the packages. This would be more like a pkgs/build-support to me.

Comment (garbas) [inline lib/update.nix]: :+1:

File: pkgs/gecko/default.nix [hunk 2]
```diff
@@ -86,4 +99,10 @@ stdenv.mkDerivation {
   shellHook = ''
     export MOZBUILD_STATE_PATH=$PWD/.mozbuild
   '';
+  passthru.update_src = updateFromGitHub {
```
Comment (zimbatm) [inline pkgs/gecko/default.nix]: `updateSrc`?

File: update.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  packages = if pkg == null
+    then pkgs_mozilla.lib.packagesToUpdate
+    else [(builtins.getAttr pkg pkgs_mozilla).update_src];
+in pkgs_mozilla.nixpkgs.stdenv.mkDerivation {
+  name = "update-nixpkgs-mozilla";
+  buildCommand = ''
+    echo "+--------------------------------------------------------+"
+    echo "| Not possible to update repositories using \`nix-build\`. |"
+    echo "|         Please run \`nix-shell update.nix\`.             |"
+    echo "+--------------------------------------------------------+"
```
Comment (zimbatm) [inline update.nix]: add `exit 1`?

**Comment by zimbatm** [comment]
Really cool overall

File: update.nix (hunk 3 above)
Comment (garbas) [inline update.nix]: :+1: will do

**Comment by garbas** [comment]
@zimbatm tnx! i guess next step would be to hook this somehow with hydra/travis.

**Comment by zimbatm** [comment]
Is it possible to request a machine from Mozilla to install the hydra instance? I like Travis in general but haven't found it to be the best experience for nix.

**Comment by garbas** [comment]
@zimbatm for the time being we can use my instance of hydra (http://hydra.garbas.si)

but how would we

- run ``nix-shell update.nix``
- then ``nix-build release.nix -A release``
- and finally automatically merge when build succeeds

in hydra?

**Comment by garbas** [comment]
maybe we need a combination of travis and hydra.

**Comment by garbas** [comment]
hmm actually to be more "mozilla friendly" we might consider [taskcluster](https://taskcluster.net).

**Comment by garbas** [comment]
we could use [taskcluster hooks](https://tools.taskcluster.net/hooks/#releng/pgo) for periodic runs of ``nix-shell update.nix`` and merge everything to ``next`` branch

then hydra would pick up new stuff from ``next`` branch an build it

we can then use taskcluster hooks to run another task to check if next branch is built successfully and autocommit to master or at least open pull request.

**Comment by zimbatm** [comment]
We could start by having taskcluster pushing the new hashes to master and then have hydra advance it's channel on success.

**Comment by garbas** [comment]
cron job is now enabled via travis it will run daily. still nothing gets automatically committed.

since we have only few packages i will just try to build all packages in travis. if succeeds i'll commit it to master, if not i will open an issue pointing to failed build and attached the diff. once changes land on master branch they will get build by hydra (http://hydra.garbas.si)

it is not perfect, but for now it should be ok.

**Pull request merged**